File: ObjC/AoC2025/AoC2025/Utility/AOCInput.m
//
//  AOCInput.m
//  AoC2017
//
//  Created by Simon Biickert on 2023-01-27.
//

#import <Foundation/Foundation.h>
#import "AOCInput.h"

NSString * const INPUT_FOLDER = @"~/Developer/Advent of Code/2025/AdventOfCode2025/Input";

@implementation AOCInput

+ (NSArray<AOCInput *> *)getInputsFor:(AOCSolution *)solution {
	NSMutableArray<AOCInput *> *inputs = [NSMutableArray array];
	
	// Challenge Input
	NSString *name = [AOCInput makeFilenameForDay:solution.day isTest:false];
	[inputs addObject: [[AOCInput alloc] initWithSolution:solution
												 filename:name
												  atIndex:0]];
	
	name = [AOCInput makeFilenameForDay:solution.day isTest:true];
	if (solution.emptyLinesIndicateMultipleInputs) {
		NSArray<NSArray<NSString *> *> *gInput = [AOCInput readGroupedInputFile:name];
		for (int i = 0; i < gInput.count; i++) {
			[inputs addObject:[[AOCInput alloc] initWithSolution:solution filename:name atIndex:i]];
		}
	}
	else {
		[inputs addObject:[[AOCInput alloc] initWithSolution:solution filename:name atIndex:0]];
	}
	
	return inputs;
}

+ (NSArray<AOCInput *> *)getTestsForSolution:(AOCSolution *)solution {
	NSArray<AOCInput *> *inputs = [AOCInput getInputsFor:solution];
	NSMutableArray<AOCInput *> *tests = [NSMutableArray array];
	
	for (int i = 1; i < inputs.count; i++) {
		[tests addObject: [inputs objectAtIndex:i]];
	}
	return tests;
}

+ (AOCInput *)getChallengeForSolution:(AOCSolution *)solution {
	return [[AOCInput getInputsFor:solution] objectAtIndex:0];
}


+ (NSString *)makeFilenameForDay:(int)day isTest:(BOOL)isTest {
	NSString *label = isTest ? @"test" : @"challenge";
	return [NSString stringWithFormat:@"day%02d_%@.txt", day, label];
}

+ (NSURL *)getInputPathFor:(NSString *)filename {
	NSString *inputFolder = [INPUT_FOLDER stringByExpandingTildeInPath];
	NSURL *folderPath = [NSURL fileURLWithPath:inputFolder isDirectory:YES];
	NSURL *filePath = [folderPath URLByAppendingPathComponent:filename];
	return filePath;
}

+ (NSArray<NSString *> *)readInputFile:(NSString *)name
					removingEmptyLines:(BOOL)removeEmptyLines {
	NSURL *url = [AOCInput getInputPathFor:name];
	
	NSError *error = nil;
	NSString *input = [NSString stringWithContentsOfFile:url.path
												encoding:NSUTF8StringEncoding
												   error:&error];
	if (error != nil) {
		NSLog(@"Error reading file %@: %@", name, error);
		return nil;
	}
	
	NSArray<NSString *> *results = [input componentsSeparatedByString:@"\n"];
	
	if (removeEmptyLines) {
		NSPredicate *p = [NSPredicate predicateWithFormat:@"SELF.length > 0"];
		return [results filteredArrayUsingPredicate:p];
	}
	return results;
}

+ (NSArray<NSString *> *)readGroupedInputFile:(NSString *)name atIndex:(int)index {
	NSArray<NSArray<NSString *> *> *groups = [AOCInput readGroupedInputFile:name];
	
	if (index < 0 || index >= groups.count) {
		return [NSArray array];
	}
	return [groups objectAtIndex:index];
}

+ (NSArray<NSArray<NSString *> *> *)readGroupedInputFile:(NSString *)name {
	NSMutableArray<NSArray<NSString *> *> *results = [NSMutableArray array];
	NSArray<NSString *> *lines = [AOCInput readInputFile:name removingEmptyLines:NO];
	
	NSMutableArray<NSString *> *group = [NSMutableArray array];
	for (NSString *line in lines) {
		if (line.length > 0) {
			[group addObject:line];
		}
		else {
			[results addObject:group];
			group = [NSMutableArray array];
		}
	}
	
	if (group.count > 0) {
		[results addObject:group];
	}
	
	return results;
}

- (AOCInput *)initWithSolution:(AOCSolution *)solution
					  filename:(NSString *)filename
					   atIndex:(int)index {
	self = [super init];
	
	_solution = solution;
	_filename = filename;
	_index = index;
	
	return self;
}

- (NSString *)getID {
	return [NSString stringWithFormat:@"%@[%d]", self.filename, self.index];
}

@end

File: Swift/Advent of Code 2024.swiftpm/Libraries/AoCInput.swift
//
//  AoCInput.swift
//  AoC 2023
//
//  Created by Simon Biickert on 2023-11-27.
//

import Foundation

struct AoCInput {
    public static var YEAR = 2023
    
    public static func fileName(day: Int, isTest: Bool) -> String {
        return "\(String(format: "day%02d", arguments: [day]))_\(isTest ? "test" : "challenge")"
    }
    
    public static func inputPath(for fileName: String) -> URL {
        let filePath = Bundle.main.path(forResource: fileName, ofType: "txt") ?? ""
        //print(filePath)
        return URL(fileURLWithPath: filePath)
        //return URL(filePath: filePath, directoryHint: .checkFileSystem, relativeTo: nil)
    }
    
    public static func inputsFor(solution s: AoCSolution) -> [AoCInput] {
        var result = [AoCInput]()
        // Challenge input. N = 1
        let challengeFile = fileName(day: s.day, isTest: false)
        result.append(AoCInput(solution: s, fileName: challengeFile, index: 0))
        
        // Test input. N = 0..N
        if s.emptyLinesIndicateMultipleInputs {
            let testFile = AoCInput.fileName(day: s.day, isTest: true)
            let testGroups = AoCInput.readGroupedInputFile(
                named: testFile)
            for (i, _) in testGroups.enumerated() {
                result.append(AoCInput(solution: s, fileName: testFile, index: i))
            }
        }
        else {
            result.append(AoCInput.init(solution: s, fileName: AoCInput.fileName(day: s.day, isTest: true), index: 0))
        }
        
        return result
    }
    
    public static func readInputFile(named fileName: String, removingEmptyLines removeEmpty:Bool) -> [String] {
        var results = [String]()
        do {
            let path = AoCInput.inputPath(for: fileName)
            let data = try Data(contentsOf: path)
            if let inputStr = String(data: data, encoding: .utf8) {
                results = inputStr.components(separatedBy: "\n")
            }
            else {
                NSLog("Could not decode \(path) as UTF-8")
            }
        } catch {
            NSLog("Could not read file \(fileName)")
        }
        if removeEmpty {
            results = results.filter { $0.count > 0 }
        }
        else {
            // Will remove empty last lines, regardless: just copy and paste error
            while results.last!.isEmpty {
                let _ = results.popLast()
            }
        }
        return results
    }
    
    public static func readGroupedInputFile(named name: String, atIndex index: Int) -> [String] {
        let result = [String]()
        guard index >= 0 else { return result }
        
        let groups = readGroupedInputFile(named: name)
        guard index < groups.count else { return result }
        
        return groups[index]
    }
    
    public static func readGroupedInputFile(named name: String) -> [[String]] {
        var results = [[String]]()
        let lines = readInputFile(named: name, removingEmptyLines: false)
        
        var group = [String]()
        for line in lines {
            if line.count > 0 {
                group.append(line)
            }
            else {
                results.append(group)
                group = [String]()
            }
        }
        if group.count > 0 {
            results.append(group)
        }
        
        return results
    }
    
    let solution: AoCSolution
    let fileName: String
    let index: Int
    var id: String {
        return "\(fileName)[\(index)]"
    }
    
    public var inputPath: URL {
        return AoCInput.inputPath(for: fileName)
    }
    
    public var textLines: [String] {
        if solution.emptyLinesIndicateMultipleInputs {
            return AoCInput.readGroupedInputFile(named: fileName, atIndex: index)
        }
        return AoCInput.readInputFile(named: fileName, removingEmptyLines: false)
    }
}
